File: DynamicProgramming/Leetcode/Medium/BestTimetoBuyandSellStockwithTransactionFee/BestTimetoBuyandSellStockwithTransactionFee.cpp
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/
#include <bits/stdc++.h>
using namespace std;

// MEMOIZATION
int maximize(vector<int> &prices, int ind, int canBuy, int fee, vector<vector<int>> &dp)
{
    if (ind == prices.size())
        return 0;
    if (dp[ind][canBuy] != -1)
        return dp[ind][canBuy];

    if (canBuy)
    {
        return dp[ind][canBuy] = max(maximize(prices, ind + 1, 0, fee, dp) - prices[ind] - fee,
                                     maximize(prices, ind + 1, 1, fee, dp));
    }
    else
    {
        return dp[ind][canBuy] = max(maximize(prices, ind + 1, 1, fee, dp) + prices[ind],
                                     maximize(prices, ind + 1, 0, fee, dp));
    }
}
int maxProfit(vector<int> &prices, int fee)
{
    vector<vector<int>> dp(prices.size(), vector<int>(2, -1));
    return maximize(prices, 0, 1, fee, dp);
}

// TABULATION
int maxProfit(vector<int> &prices, int fee)
{
    vector<vector<int>> dp(prices.size() + 1, vector<int>(2, 0));
    for (int ind = prices.size() - 1; ind >= 0; ind--)
    {
        for (int canBuy = 0; canBuy <= 1; canBuy++)
        {
            if (canBuy)
            {
                dp[ind][canBuy] = max(dp[ind + 1][0] - prices[ind] - fee, dp[ind + 1][1]);
            }
            else
            {
                dp[ind][canBuy] = max(dp[ind + 1][1] + prices[ind], dp[ind + 1][0]);
            }
        }
    }

    return dp[0][1];
}

File: DynamicProgramming/Leetcode/Medium/BestTimetoBuyandSellStockwithTransactionFee/BestTimetoBuyandSellStockwithTransactionFee.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/

import java.util.Arrays;

public class BestTimetoBuyandSellStockwithTransactionFee {

    // MEMOIZATION
    int maximize(int ind, int[] prices, int fee, int buy, int[][] dp) {
        if(ind == prices.length) return 0;
        if(dp[buy][ind] != -1) return dp[buy][ind];
        int profit = 0;
        if(buy == 1) {
            profit = Math.max(-prices[ind] - fee + maximize(ind+1, prices, fee, 0, dp), maximize(ind+1, prices, fee, 1, dp));
        } else {
            profit = Math.max(prices[ind] + maximize(ind + 1, prices, fee, 1, dp), maximize(ind + 1, prices, fee, 0, dp));
        }
        dp[buy][ind] = profit;
        return profit;
    }
    public int maxProfit(int[] prices, int fee) {
        int[][] dp = new int[2][prices.length];
        Arrays.fill(dp[0], -1);
        Arrays.fill(dp[1], -1);
        return maximize(0, prices, fee, 1, dp);
    }

    // TABULATION
    public int maxProfit(int[] prices, int fee) {
        int[][] dp = new int[2][prices.length + 1];
        for(int ind=prices.length-1; ind>=0; ind--) {
            for(int buy=1; buy>=0; buy--) {
                int profit = 0;
        if(buy == 1) {
            profit = Math.max(-prices[ind] - fee + dp[0][ind+1], dp[1][ind+1]);
        } else {
            profit = Math.max(prices[ind] + dp[1][ind+1], dp[0][ind+1]);
        }
        dp[buy][ind] = profit;
            }
        }
        return dp[1][0];
    }
}